sigs/jvm/rfcs/20200126-tensorflow-data.md

```diff
@@ -0,0 +1,209 @@
+| **Current Implementation** | [dhruvrajan/tensorflow-java](https://github.com/dhruvrajan/tensorflow-java/tree/tf-data-cherrypick/tensorflow-frameworks/tensorflow-data) |
+| **Author(s)** | Dhruv Rajan (dhruv@krishnaprem.com) |
+| **Sponsor**   | Karl Lessard (karl.lessard@gmail.com)                 |
+| **Updated**   | 2020-01-26                                  |
+
+## Objective
+
+To provide simple APIs for loading data from various formats, and preparing
+datasets for use in training deep learning models with TensorFlow.
+
+This API design for Java follows that of Python's
+[tf.data](https://www.tensorflow.org/api_docs/python/tf/data).
+
+## Motivation
+
+A large part of the standard deep-learning workflow is managing datasets:
+configuring iteration and batching over a sequence of elements,
+and performing transformations on individual or grouped dataset elements.
+
+The goal of this project is to provide a high-level abstraction which simplifies
+this workfow, based on TensorFlow data ops, with syntax closely mirorring
+Python's [tf.data](https://www.tensorflow.org/api_docs/python/tf/data).
+
+## User Benefit
+
+Users of TensorFlow Java will be able to use this framework to easily
+load, transform, and iterate over datasets.
+
+Additionally this framework will be used at the core of the automated
+training process in TensorFlow Java's port of [tf.keras](https://www.tensorflow.org/api_docs/python/tf/keras).
+
+## Design Proposal
+
+The `Dataset` class represents a sequence of elements which can be iterated over and
+transformed. Each element is a list of "output" operands, represented by the type `List<Output<?>>`. 
+
+Note: An `Output` is a symbolic handle to a tensor produced by a TensorFlow op. In graph
+mode, `Output` objects will not have a concrete `Tensor` value unless all dependent operations
+are run in a `Session` (this is done "in-real-time" in eager mode).
+
+### Construction
+
+Datasets can be constructed either directly from a data source (e.g. a list of tensors representing the components of the dataset), or as a transformation on an existing dataset.
+
+#### From Data Source
+
+To construct a dataset from a list of tensor components, use 
+`Dataset.fromTensorSlices( ... )`. For example, say we are working
+with a standard feature/label dataset which has 4 elements.
+
+```java
+float[][] features = new float[][] {
+    {1, 2, 3},
+    {4, 5, 6},
+    {7, 8, 9},
+    {10, 11, 12}
+}
+
+float[] labels = new float[] {
+    0,
+    1,
+    1,
+    0
+}
+```
+
+A dataset can be constructed from a list of the constant `Operand`s generated
+from this dataset, and a list of `DataType` objects corresponding
+to the type of each component:
+
+Note: Each of the input components must share the same first "batch" dimension.
+
+```java
+Ops tf = // ... TensorFlow Ops accessor (either graph or eager)
+Dataset dataset = Dataset.fromTensorSlices(
+    Arrays.asList(tf.constant(features), tf.constant(labels)),
```

Comment:
or could we do something like:
```java
Dataset.fromTensorSlices(tf, Arrays.asList(features, labels), ...)
```
and let the `Dataset` create the constant? Also, the datatypes might became optional since often then can be inferred from the type of the constant.

Reply:
Inferring the type automatically here would be cool!

Keeping this interface  (as well) might be useful though, in case users have tensorflow operations they want to run before constructing the dataset? That'd be a bit strange, but possible, I suppose...

Reply:
Also, I think there aren't `tf.constant` signatures which allow input of `NdArray<?>`,  so it might be tough to preserve the ability to input tensors with different dtypes.

We could add multiple typed factories here, though, for `IntNdArray`, `FloatNdArray`, etc...

Reply:
mmh, there are... for example, [this one](https://github.com/tensorflow/java/blob/9fe9d87a454308e020b57839abf2e061f04c5a1e/tensorflow-core/tensorflow-core-api/src/gen/annotations/org/tensorflow/op/Ops.java#L1363). That is what you were looking for, right?